### DISCUSSION TITLE
Improving Flatten.js

### DISCUSSION THREAD

**Discussion by edemaine** [alexbol99/flatten-js#131] [General]
@alexbol99 I've started using Flatten.js in a currently small but hopefully growing [computational geometry playground](https://erikdemaine.org/compgeom/) tool ([GitHub](https://github.com/edemaine/compgeom)). As a result, I'm interesting in improving Flatten.js; it's already great, but I'm running into small things here and there (#129, #130, lack of e.g. angular bisector computation).

Would you like help maintaining the package (e.g. reviewing existing PRs), or a series of new PRs, or should I make my own fork? Let me know what's best for you! Or let me know if you'd like to meet to discuss.

**Comment by alexbol99** [comment]
Hi, Erik,
Thank you for choosing my library for your project.
It looks great, and I will be happy to collaborate with you.
Previously I also wanted to create something like this, but then I've met https://observablehq.com/ and it covered all my needs.
What is the use case of your compgeom playground?
Are you planning to make it interactive? To add animation? Support formats?
You are right, recently I did not have enough time maintenance, but I think I can give it a push.
I will start from small things, let's see how it's going on

**Reply by edemaine** [reply]
Thanks for the quick and positive response!

My original motivation was to make something like Geometer's Sketchpad, but where I had the full power of a programming language (JavaScript/CoffeeScript) instead of having to click buttons a zillion times to repeat constructions in programmatic patterns.  (Specifically, last week, I needed to draw ~100 lines, each of which required ~5 operations, but it was just a few minutes with `for` loops.)

Obviously, the project is still in its infancy (it's been only ~1 day of development so far), but I'm exciting about the potential directions it could go.  I just added a simple rectangular zoom tool so that you can look closer at specific parts of the drawing (to see what intersects what, for example), and save the resulting SVG.

It'd definitely be cool to support interactive geometry constructions, like Geometer's Sketchpad and your [ObservableHQ](https://observablehq.com/@alexbol99/inversion-in-a-circle) examples. I think there should be a much simpler interface possible than d3. I've been thinking of tracking the dependency DAG on the elements, and allowing any of those with degrees of freedom to be dragged, with automatic updating of just the affected elements.

I've also been wondering about making a Geometer's Sketchpad-like GUI interface, where you can interactively click on elements to do geometric constructions, and it adds the corresponding code to the Playground.  I'm not quite sure how this would interact with the live code editing (how does the GUI even know the correct variable name for a selected element?); maybe it would need some manual adjustment.

Let me know if you have ideas!

ObservableHQ definitely looks cool; thanks for the pointer. One obvious advantage of this project is that it's completely open source, so it's easy for me or anyone to add features that they want. And what I have in mind is much more tuned to displaying geometry.
